### PR TITLE
CI to verify that apps can be built on every push/PR

### PR DESCRIPTION
**Pull request by JuanSapriza**
# The problem
Changes done on HW or SW may affect the health of applications in the repository (e.g. renaming a register in hw and not updating every occurrence in sw may lead to applications not compiling). 

# The proposed solution
Have an Github Action executed on every push and PR that builds (`make app`) every application inside the `sw/applications` folder and makes sure they can all be built. 
This is not an extensive test, as applications are not simulated. This guarantees compilation, not proper functionality. 
Testing the functionality is a more ambitious project requiring all applications to return or report their status. 

# The implementation
To perform actions, Github hosts `runners` (servers running either Linux, Windows or Mac OSs). 
Runners look for _actions_ as `*.yml` files inside the `.github/workflows` directory of the repository. 
Each action can define several _jobs_, i.e. a set of _steps_ to be run on a specified _container_. 
To prevent installing all necessary requirements every time, the container is built from a prepared _image_ uploaded to my personal Github repo as a _package_: [xheep-compiler:riscv-clang](https://github.com/users/JuanSapriza/packages/container/package/xheep-compiler). 
That package was commited from a container running `ubuntu:focal` (20.04), in which the RISCV compiler and Clang were installed. 
Python requirements were not pre-installed as they drastically increased the size of the image and the time consumed to load the image ended up being larger than the one required to install them every time once the container was up and running on Github's servers.  

Once the job is launched and the container image is built and started, the action runs a script that is located inside the `.github/workflows/build-apps-job` directory. It is in charge of installing the required python modules and launching a second script inside the same folder. 
The script `build-apps.sh` is the one in charge of _make apping_ all applications both with _gcc_ and _clang_ and reporting the success or failure of the procedure. *This script can also be run locally to generate the same output that the Github action would.*

Only if **ALL** applications were successfully built for **BOTH** compilers, the action is considered **PASSED**.
The check takes around 3 minutes to run.

## If the check is not passed
If a PR has checks that do not pass, you can follow the instructions from inside the PR to get some insight on the errors that caused the failure. 

Action diagram

## Other changes
In order to submit a CI that actually passes itself, some changes were done to the rest of the repository:
* `SPI_START_ADDRESS` was changed throughout the sw to `SPI2_START_ADDRESS`.
* The instruction memory was increased to `0x00010000` inside the `mcu_cfg.hjson` to fit _freeRTOS_ when compiled with clang. 
* The `CMakeLists.txt` was modified to include the _freeRTOS_ linker files when compiled with clang. 

## Future work
* Move the package to ESL Github
* Add documentation to allow people to use the package to build applications for `X-HEEP`
* Offer more and more complete packages (make one plug-n-play)

### CODE REVIEW

**Comment by JoseCalero** [comment]
Please, update the PR with the last changes in main :)

**Comment by JuanSapriza** [comment]
@JoseCalero this should be good to go now

File: .github/workflows/building.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+name: Building tests
+on: [push, pull_request]
+
+jobs:
+  test_applications:
+    runs-on: ubuntu-latest
+    container: 
+      image: ghcr.io/juansapriza/xheep-compiler:riscv-clang
```
Comment (JoseCalero) [inline .github/workflows/building.yml:8]: we should be having the image using the lab user rather than a personal one

Comment (JoseCalero) [reply .github/workflows/building.yml:8]: @JuanSapriza any news about this?

Comment (JuanSapriza) [reply .github/workflows/building.yml:8]: not yet :(

**Comment by JuanSapriza** [comment]
I do not understand how... but now that we host the container at ESL, the building of some apps fail BUT ONLY WHEN THE CI IS RAN IN THE ESL REPO. On my personal repo, the same commit passes the tests. 

Screenshot from 2023-05-12 11-20-00
The last commit, tested on my repo.

Screenshot from 2023-05-12 11-20-05
The last commit tested on this repo.

**Comment by JuanSapriza** [comment]
These are the apps that failed the tests. Now both my repository and this one are consistent. 
Screenshot from 2023-05-12 12-03-57

Details can be obtained here. 
Screenshot from 2023-05-12 12-04-59

**Comment by JoseCalero** [comment]
you can merge main and this should be good to go

**Comment by JuanSapriza** [comment]
All checks passed. Ready to merge @JoseCalero

**Pull request merged**